packages/@aws-cdk/core/lib/private/cloudformation-lang.ts

```diff
@@ -43,8 +43,8 @@ export class CloudFormationLang {
 
     // Some case analysis to produce minimal expressions
     if (parts.length === 1) { return parts[0]; }
-    if (parts.length === 2 && typeof parts[0] === 'string' && typeof parts[1] === 'string') {
-      return parts[0] + parts[1];
+    if (parts.length === 2 && ['string', 'number'].includes(typeof parts[0]) && ['string', 'number'].includes(typeof parts[1])) {
```

Comment:
Can we use ` isConcatable()` here? Or can these actually be Tokens themselves? If so, do we need a different helper?